src/db.c

```diff
@@ -511,7 +537,7 @@ int findSlotByKeyIndex(redisDb *db, unsigned long target, unsigned dictType) {
 int dbGenericDelete(redisDb *db, robj *key, int async, int flags) {
     dictEntry **plink;
     int table;
-    int slot = getKeySlot(key->ptr);
+    int slot = (flags & DB_FLAG_KEY_EVICTED) ? calculateKeySlot(key->ptr) : getKeySlot(key->ptr);
```

Comment:
that's a bit awkward, this tip was created for a different reason, and the assumption that only in this flow we have to calculate the slot rather than use the cached one seems odd and dangerous.
at the very least, there should be a big comment here, but i'd rather find another solution.
let's discuss.

Reply:
The other solution which I thought about was we could provide the slot information which is already available in the eviction code flow but would require us to introduce another method or alter the parameters in this API and affect all other callers.

Reply:
Maybe inside of calculateKeySlot() we should be evaluating whether or not we're in the main command execution part (i.e., inside the call, I think the flag is `CLIENT_EXECUTING_COMMAND`). The slot is only valid during the command execution, not during other side effects.

Reply:
i thought of that too, but i was also afraid that some side effect mechanism will attempt to access other keys from within call as well, like eviction does.
i suppose we don't have these in redis (e.g. no eviction in EXEC or EVAL, and i can't think of anything else like that).
we do have both scripts and modules accessing random keys inside commands, and these should be handled separately (please ack).
so bottom line, i support your suggestion, please also add some comments around that.